### PR TITLE
#5804 Versioning doesn't work with a custom type primary key

### LINKED ISSUE #5804: Versioning doesn't work with a custom type primary key
I've been wondering for a while why does Doctrine resets my `@Version` entity field to `0` every time I flush to the database even though the correct version can be found in the database, so I decided to dig down into the entrails of Doctrine and found some peculiar issue.

When the execution flow reaches the `Doctrine\ORM\Persisters\Entity\BasicEntityPersister::fetchVersionValue` method (it's the point after the flush where Doctrine tries to retrieve the new version of the entity from the database to re-hydrate the entity) apparently it assumes that the primary key for this table is a known primitive type and doesn't bother trying to map it to anything. More specifically this line:

``` php
// Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:346
$value = $this->conn->fetchColumn($sql, array_values($flatId));
```

The third and fourth parameters of `fetchColumn` should be `0` and the type(s) of the `$flatId` field(s). But nothing is sent. In my situation, my primary key is a custom type `uuid_binary`. Then down the line when you get to `Doctrine\DBAL\Connection::executeQuery` and the statement is being prepared, since there is no types involved in the `$types` parameter, the binding actually never happens!

``` php
// Doctrine/DBAL/Driver/PDOStatement.php:825
$stmt = $this->_conn->prepare($query);
if ($types) {
    $this->_bindTypedValues($stmt, $params, $types);
    $stmt->execute();
} else {
    $stmt->execute($params);
}
```

Which means that the versioning query...

``` sql
SELECT version FROM tablename WHERE primarykey = ?
```

... doesn't ever receive any value for the `primarykey` placeholder and the request ultimately fails and returns `false`. Doctrine then takes this `false` and assigns it to the `@Version` field of the entity in `Doctrine\ORM\Persisters\Entity\BasicEntityPersister::assignDefaultVersionValue`, which then translates to `0`.

``` php
// Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:317
protected function assignDefaultVersionValue($entity, array $id)
{
    $value = $this->fetchVersionValue($this->class, $id);
    $this->class->setFieldValue($entity, $this->class->versionField, $value);
}
```

So it seems like this scenario is entirely missing.

### PR DESCRIPTION
**Pull request by alle**
This PR fixes #5804.
It's not a my own solution, but the one provided by @ngrevet-lightspeed.
Since I need this fixed, I made this PR on behalf of @ngrevet-lightspeed.

### CODE REVIEW

**Review by Ocramius — CHANGES_REQUESTED** [review]
@alle the patch needs tests to be validated

**Comment by alle** [comment]
@Ocramius @ngrevet-lightspeed Everything should be fine, now. I was too hurried, last time. Sorry.

**Comment by Ocramius** [comment]
@alle I still don't see test changes though :-\

**Comment by alle** [comment]
@Ocramius Well, I thought tests shouldn't change, should they?
I thought I fixed this bug, working on the `BasicEntityPersister::fetchVersionValue()` method, by passing the involved tests, without changing them. Humbly, am I missing something?

**Comment by Ocramius** [comment]
@alle a test demonstrating the original bug (failing with the original repository state, passing with this patch) is needed ;-)

**Comment by alle** [comment]
@ocramius: @astepin did the job (in #6496) that I should have done from myself.
@astepin The only change I made is in the custom type `getSQLDeclaration()` method: your version was failing against the Postgres platform, since Doctrine maps to a specific datatype for that platform.

**Comment by alle** [comment]
I changed the involved classes in tests, complying to other tests' structure, in order to avoid a merge conflict on the `SchemaValidatorTest`, due to recent updates

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 1]
```diff
@@ -344,9 +344,16 @@ protected function fetchVersionValue($versionedClass, array $id)
              . ' FROM '  . $tableName
              . ' WHERE ' . implode(' = ? AND ', $identifier) . ' = ?';
 
+        $types = [];
+        foreach ($id as $field => $value) {
+            foreach ($this->getTypes($field, $value, $versionedClass) as $type) {
```
Comment (lcobucci) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]: Do we need to iterate this too? I mean in theory we could use `array_merge()` right, like:

```php
$types = [];

foreach ($id as $field => $value) {
    $types = array_merge($types, $this->getTypes($field, $value, $versionedClass));
}
```

Comment (alle) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]: @lcobucci You're simply right.

**Comment by alle** [comment]
@Ocramius @lcobucci Is there any possibility to have this PR merged? I didn't see any news since the last changes.

File: tests/Doctrine/Tests/Models/GH5804/GH5804Article.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * @Table(name="gh5804_articles")
+ */
+class GH5804Article
+{
+    /**
+     * @Id
+     * @Column(type="GH5804Type")
+     * @GeneratedValue(strategy="CUSTOM")
+     * @CustomIdGenerator(class="\Doctrine\Tests\Models\GH5804\GH5804Generator")
+    ")
```
Comment (alcaeus) [inline tests/Doctrine/Tests/Models/GH5804/GH5804Article.php]: Syntax error ;)

File: tests/Doctrine/Tests/Models/GH5804/GH5804Article.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * @Entity
+ * @Table(name="gh5804_articles")
+ */
+class GH5804Article
+{
+    /**
+     * @Id
+     * @Column(type="GH5804Type")
+     * @GeneratedValue(strategy="CUSTOM")
+     * @CustomIdGenerator(class="\Doctrine\Tests\Models\GH5804\GH5804Generator")
```
Comment (alcaeus) [inline tests/Doctrine/Tests/Models/GH5804/GH5804Article.php]: Could/Should use `::class` constant

File: tests/Doctrine/Tests/Models/GH5804/GH5804Article.php (hunk 2 above)
Comment (alle) [inline tests/Doctrine/Tests/Models/GH5804/GH5804Article.php]: Aaargh! Didn't see that!

File: tests/Doctrine/Tests/Models/GH5804/GH5804Article.php (hunk 3 above)
Comment (alle) [inline tests/Doctrine/Tests/Models/GH5804/GH5804Article.php]: In which form? Any example?

Comment (alcaeus) [inline tests/Doctrine/Tests/Models/GH5804/GH5804Article.php]: Paired with a `use` statement: `@CustomIdGenerator(class=GH5804Generator::class)`. Works with Annotations >= 1.2 IIRC.

Comment (alle) [inline tests/Doctrine/Tests/Models/GH5804/GH5804Article.php]: I guessed it was that way, but my local tests fail. I even update merged with the latest `master` release, updated the dependencies (my Annotations version was right, anyway), but your suggestion (which seems to be logical) doesn't work to me. Did you try by yourself?
Anyway, even without that `::class` notation, everything is fine.

Comment (alcaeus) [inline tests/Doctrine/Tests/Models/GH5804/GH5804Article.php]: That's strange. I've been using that for a couple of years now.

**Comment by lcobucci** [comment]
I'll take a look at this in a bit

File: tests/Doctrine/Tests/Models/GH5804/GH5804Article.php (hunk 3 above)
Comment (lcobucci) [inline tests/Doctrine/Tests/Models/GH5804/GH5804Article.php]: It requires the FQCN, just that 😉

**Comment by lcobucci** [comment]
@alle thanks for your contribution, I've simplified things a bit (kept all test related classes together).

@Ocramius it LGTM, can you check this one to see if we're good to go?

**Review by lcobucci — APPROVED** [review]

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 4]
```diff
@@ -344,9 +344,14 @@ protected function fetchVersionValue($versionedClass, array $id)
              . ' FROM '  . $tableName
              . ' WHERE ' . implode(' = ? AND ', $identifier) . ' = ?';
 
+        $types = [];
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php]: Can this block be moved to a private method?

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH5804Test.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+class GH5804Article
+{
+    /**
+     * @Id
+     * @Column(type="GH5804Type")
+     * @GeneratedValue(strategy="CUSTOM")
+     * @CustomIdGenerator(class=\Doctrine\Tests\ORM\Functional\Ticket\GH5804Generator::class)
+     */
+    public $id;
+    /**
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH5804Test.php]: Spacing

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH5804Test.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            return null;
+        }
+
+        return 'testGh5804DbValue';
+    }
+}
+
+/**
+ * @Entity
+ * @Table(name="gh5804_articles")
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH5804Test.php]: Not really needed

**Review by Ocramius — APPROVED** [review]
Some nitpicks apply

**Comment by lcobucci** [comment]
@alle 🚢 thanks!

**Comment by alle** [comment]
Thank you so much, guys!
This was my first opportunity to contribute to this project, even to "calibrate" myself to future contributions. Or at least I hope so!

**Pull request merged**